improver/regrid/landsea2.py

```diff
@@ -0,0 +1,212 @@
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""
+Plugin to regrid using custom nearest and bilinear methods, both with
+land-sea awareness
+"""
+
+import numpy as np
+
+from improver import BasePlugin
+from improver.regrid.bilinear import (
+    adjust_for_surface_mismatch,
+    apply_weights,
+    basic_indexes,
+    basic_weights,
+)
+from improver.regrid.grid import (
+    classify_input_surface_type,
+    classify_output_surface_type,
+    convert_from_projection_to_latlons,
+    create_regrid_cube,
+    flatten_spatial_dimensions,
+    latlon_from_cube,
+    slice_cube_by_domain,
+    slice_mask_cube_by_domain,
+    unflatten_spatial_dimensions,
+)
+from improver.regrid.idw import similar_surface_classify
+from improver.regrid.nearest import nearest_regrid, nearest_with_mask_regrid
+
+NEAREST = "nearest"
+BILINEAR = "bilinear"
+WITH_MASK = "-with-mask"
+BILINEAR2 = f"{BILINEAR}-2"
+NEAREST2 = f"{NEAREST}-2"
+NEAREST_MASK2 = f"{NEAREST}{WITH_MASK}-2"
+BILINEAR_MASK2 = f"{BILINEAR}{WITH_MASK}-2"
+NUM_NEIGHBOURS = 4
+
+
+class RegridWithLandSeaMask(BasePlugin):
+    """
+    Replace data values at points where the nearest-regridding technique
+    selects a source grid-point with an opposite land-sea-mask value to the
+    target grid-point.
+    The replacement data values are selected from a vicinity of points on the
+    source-grid and the closest point of the correct mask is used.
+    Where no match is found within the vicinity, the data value is not changed.
+    Note: regrid_mode options are "nearest-2", "nearest-with-mask-2","bilinear-2", 
+    and "bilinear-with-mask-2" in this class.
+    
+    """
+
+    def __init__(self, regrid_mode="bilinear-2", vicinity_radius=25000.0):
+        """
+        Initialise class
```

Comment:
I think you need a docstring here to explain what the arguments are.

Reply:
added